Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs

```diff
@@ -0,0 +1,631 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+// #defines for features that are not yet implemented
+#define TODO_ROTATE_FACE_USER
+
+using MixedRealityToolkit.Common;
+using MixedRealityToolkit.Common.Extensions;
+using MixedRealityToolkit.InputModule.EventData;
+using MixedRealityToolkit.InputModule.Focus;
+using MixedRealityToolkit.InputModule.InputHandlers;
+using MixedRealityToolkit.InputModule.InputSources;
+using MixedRealityToolkit.UX.BoundingBoxes;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+using UnityEngine.Assertions;
+
+
+namespace MixedRealityToolkit.InputModule.Utilities.Interations
+{
+    /// <summary>
+    /// This script allows for an object to be movable, scalable, and rotatable with one or two hands. 
+    /// You may also configure the script on only enable certain manipulations. The script works with 
+    /// both HoloLens' gesture input and immersive headset's motion controller input.
+    /// See Assets/MixedRealityToolkit-Examples/Input/Readme/README_TwoHandManipulationTest.md
+    /// for instructions on how to use the script.
+    /// </summary>
+    public class TwoHandManipulatable : MonoBehaviour, IInputHandler, ISourceStateHandler
+    {
+        // Event that gets raised when user begins manipulating the object
+        public event Action StartedManipulating;
```

Comment:
Should we replace this with calls to the InputManager.RaiseManipulationStarted/Ended?

Reply:
Hm, based on my reading of `CustomInputSource.cs` it seems like manipulation events are fired whenever a motion controller is moved and a button is down, regardless of whether something is actually being manipulated. Further, the events contain information about the controller id attached to the manipulation. In contrast the `StartedManipulating` and `StoppedManipulating` events are specific to this object--the event communicates "a specific object began being manipulated". Additionally, an object may be manipulated by more than one input source, which would not fit with the current structure of the `ManipulationEventData`. For these two reasons, I think it's better to keep the events as separate.

Reply:
The pattern I emulated was the `StartedDragging` and `StoppedDragging` events in `HandDraggable.cs`

Reply:
It doesn't matter what the input source is, we can take advantage of these events here.

Reply:
What about this example: There's some code that listens for ManipulationStarted and ManipulationEnded, assuming that this event gets fired for each controller when the controller starts dragging. Then the person adds a two handed manip component to their scene and is surprised to see a new ManipulationStarted event firing whenever the two hand manipulatable starts moving, in addition to the two ManipulationStarted events from each controller? 

Additionally, wouldn't this break the pattern that HandDraggable already has? A developer might be confused about why OneHandDraggable exposes a StartedDragging event, but TwoHandManipulatable does not, and instead for that one the dev needs to implement IManipulatable, or register events listeners in their code.

Reply:
Okay, I'll look at that this weekend.